File: scenarios/browser-automation-studio/api/automation/playwright-driver/src/utils/logger.ts
import winston from 'winston';
import type { Config } from '../config';

export function createLogger(config: Config): winston.Logger {
  const format =
    config.logging.format === 'json'
      ? winston.format.combine(
          winston.format.timestamp(),
          winston.format.errors({ stack: true }),
          winston.format.json()
        )
      : winston.format.combine(
          winston.format.timestamp(),
          winston.format.colorize(),
          winston.format.errors({ stack: true }),
          winston.format.printf(
            ({ level, message, timestamp, ...meta }) =>
              `${timestamp as string} [${level}]: ${message as string} ${
                Object.keys(meta).length > 0 ? JSON.stringify(meta) : ''
              }`
          )
        );

  return winston.createLogger({
    level: config.logging.level,
    format,
    transports: [new winston.transports.Console()],
  });
}

// Default logger instance (can be replaced via setLogger)
let loggerInstance: winston.Logger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.errors({ stack: true }),
    winston.format.json()
  ),
  transports: [new winston.transports.Console()],
});

export const logger = loggerInstance;

export function setLogger(newLogger: winston.Logger): void {
  loggerInstance = newLogger;
}

File: scenarios/prompt-injection-arena/api/logger.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"time"
)

// LogLevel represents the severity of a log message
type LogLevel string

const (
	LogLevelDebug LogLevel = "DEBUG"
	LogLevelInfo  LogLevel = "INFO"
	LogLevelWarn  LogLevel = "WARN"
	LogLevelError LogLevel = "ERROR"
	LogLevelFatal LogLevel = "FATAL"
)

// Logger provides structured logging capabilities
type Logger struct {
	service string
}

// LogEntry represents a structured log entry
type LogEntry struct {
	Timestamp string                 `json:"timestamp"`
	Level     string                 `json:"level"`
	Service   string                 `json:"service"`
	Message   string                 `json:"message"`
	Fields    map[string]interface{} `json:"fields,omitempty"`
}

// NewLogger creates a new structured logger
func NewLogger(service string) *Logger {
	return &Logger{service: service}
}

// logWithLevel logs a message at the specified level
func (l *Logger) logWithLevel(level LogLevel, message string, fields map[string]interface{}) {
	entry := LogEntry{
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Level:     string(level),
		Service:   l.service,
		Message:   message,
		Fields:    fields,
	}

	jsonBytes, err := json.Marshal(entry)
	if err != nil {
		// Fallback to simple logging if JSON marshaling fails
		log.Printf("[%s] %s: %s", level, l.service, message)
		return
	}

	fmt.Fprintln(os.Stdout, string(jsonBytes))
}

// Info logs an informational message
func (l *Logger) Info(message string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.logWithLevel(LogLevelInfo, message, f)
}

// Warn logs a warning message
func (l *Logger) Warn(message string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.logWithLevel(LogLevelWarn, message, f)
}

// Error logs an error message
func (l *Logger) Error(message string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.logWithLevel(LogLevelError, message, f)
}

// Fatal logs a fatal error and exits
func (l *Logger) Fatal(message string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.logWithLevel(LogLevelFatal, message, f)
	os.Exit(1)
}

// Debug logs a debug message
func (l *Logger) Debug(message string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	l.logWithLevel(LogLevelDebug, message, f)
}

// Global logger instance
var logger *Logger

// InitLogger initializes the global logger
func InitLogger() {
	logger = NewLogger("prompt-injection-arena-api")
}
